This demonstrates a speedup in confirmation rate during initial boots…

clemahieu opened the pull request:
…trapping due to conservative limiters in the confirmation_solicitor. Rather than limiting with a constant number, it limits based on feedback directly from network channels.

Rather than limiting on a per-channel basis, a lot of complexity can be removed with by limiting the total number of active elections based on feedback from the confirmation rate.

MajorChump commented:
This looks like the issue I found with announcements being dropped due to max_channel_requests being exceeded. 

Looking at that PR it removes max_channel_requests from the flow, probably could clean the code up a little by removing 

`L60 auto const max_channel_requests (config.confirm_req_batches_max * nano::network::confirm_req_hashes_max);`

And confirm_req_batches_max could be removed from config I think.

Reference: https://github.com/nanocurrency/nano-node/issues/3574#issuecomment-982728204

clemahieu committed:
This change limits confirmation requests by the status of the tcp channel rather than by a fixed number.

clemahieu committed:
Removing confirm_req_batches_max as it's unused.

dsiganos commented on nano/node/transport/tcp.hpp:
```diff
@@ -68,6 +68,16 @@ namespace transport
+			bool result = true;
+			if (auto socket_l = socket.lock ())
+			{
+				result = socket_l->max ();
```
This function is called full() but it calls a socket function max().
Confusingly, there is a socket called full(), which the channel function full() does not call.
I think this should be at least called max() to match the socket max().
Although, I think it would be better to call both max functions something else, half_full() maybe?

I wonder if it would be better to have full(), half_full() and quarter_full() consistent across tcp sockets and tcp channels.

dsiganos commented on nano/node/confirmation_solicitor.cpp:
```diff
@@ -70,7 +69,7 @@ bool nano::confirmation_solicitor::add (nano::election const & election_a)
 		{
 			auto & request_queue (requests[rep.channel]);
-			if (request_queue.size () < max_channel_requests)
+			if (!rep.channel->full ())
```
Completely filling up the buffers is possibly not a good idea. It might starve other processes of network bandwidth. We are going from a value of 2 to thousands. How about half or quarter full?

dsiganos committed:
Rename channel::full() to channel::max() to match socket::max()

channel::full() was calling channel::max() instead of socket::full(),
which also exists. So it makes sense for consistency to rename it.

thsfs commented:
Looks good to me, but @dsiganos suggestions make sense, it is worth to improve the function names he pointed out.
